### PR TITLE
Fix for #2272 - use of default enum method parameter throws exception

### PR DESCRIPTION
**Pull request by daveaglick**
This PR should fix issue #2272

### CODE REVIEW

**Comment by daveaglick** [comment]
So I see that there are failed tests, and they look related to my change, but I'm a little confused because I only changed code in the scripting library and the tests look like they're on more primary parts of the code. Does the test scaffolding use the scripting support?

**Comment by KevinH-MS** [comment]
@daveaglick we try to run as many of our compiler emit tests as possible through both emit code paths. These are "regular" emit and "refemit" (used by scripting).  If you change the refemit codepath, you can end up breaking the "regular" compiler tests (you'll see "RefEmit" on the callstack).

**Comment by daveaglick** [comment]
@KevinH-MS Gotcha - it looks like that's exactly what's going on.

I might need some assistance from someone who understands emit better than I do. I was able to replicate the test failure and traced it back to the call that I changed (which is `Enum.GetUnderlyingType()`). It looks like this call is throwing a `NotSupportedException: The invoked member is not supported before the type is created.` with the following stack trace:
```
at System.Reflection.Emit.TypeBuilder.GetFields(BindingFlags bindingAttr)
   at System.Type.GetEnumUnderlyingType()
   at Microsoft.CodeAnalysis.Scripting.Emit.ReflectionEmitter.DefineParameter(MethodBuilder methodBuilder, ConstructorBuilder constructorBuilder, IParameterDefinition paramDef) in e:\Code\Roslyn\src\Scripting\Core\Emit\ReflectionEmitter.cs:line 2040
   at Microsoft.CodeAnalysis.Scripting.Emit.ReflectionEmitter.DefineParameters(MethodBuilder methodBuilder, ConstructorBuilder constructorBuilder, IMethodDefinition methodDef) in e:\Code\Roslyn\src\Scripting\Core\Emit\ReflectionEmitter.cs:line 1981
   at Microsoft.CodeAnalysis.Scripting.Emit.ReflectionEmitter.EmitWorker(IMethodReference entryPoint, CancellationToken cancellationToken) in e:\Code\Roslyn\src\Scripting\Core\Emit\ReflectionEmitter.cs:line 1214
   at Microsoft.CodeAnalysis.Scripting.Emit.ReflectionEmitter.Emit(EmitContext context, IEnumerable`1 referencedAssemblies, ModuleBuilder builder, AssemblyLoader assemblyLoader, IMethodReference entryPoint, CancellationToken cancellationToken) in e:\Code\Roslyn\src\Scripting\Core\Emit\ReflectionEmitter.cs:line 1147
   at Microsoft.CodeAnalysis.Scripting.Emit.CommonCompilationExtensions.Emit(Compilation compilation, ModuleBuilder moduleBuilder, AssemblyLoader assemblyLoader, Func`2 assemblySymbolMapper, Boolean recoverOnError, DiagnosticBag diagnostics, CancellationToken cancellationToken, MethodInfo& entryPoint, Byte[]& compiledAssemblyImage) in e:\Code\Roslyn\src\Scripting\Core\Emit\CommonCompilationExtensions.cs:line 113
```

This is where my knowledge breaks down. The exception is clearly coming from `System.Reflection.Emit.TypeBuilder.GetFields(BindingFlags)` and MSDN does document that this is a possible exception for that method. However, I have no idea why this would be a "incomplete type" at this stage. I'll keep digging, but perhaps someone knows what's going on off the top of their head?

**Comment by daveaglick** [comment]
A little more info, FWIW:

I can see that `paramType` is actually a `TypeBuilder`. Also, if you call `TypeBuilder.IsCreated()` at the point where I made the change, it returns false (because `TypeBuilder.m_hasBeenCreated` is false, and it only gets set to true on `TypeBuilder.CreateType()` or `TypeBuilder.CreateTypeInfo()`, which I guess finishes the type). That's the source of the exception on my change: `Type.GetEnumUnderlyingType()` calls `TypeBuilder.GetFields(BindingFlags)` which checks `TypeBuilder.IsCreated()`.

If I undo the change and go back to `paramType.UnderlyingSystemType` it works in most cases, including the tests. However, in some cases such as the one in issue #2272, this call returns the actual enum type instead of the underlying type. This seems to happen because of the code in `TypeBuilder.UnderlyingSystemType` which has the following checks:
```
if (this.m_bakedRuntimeType != (RuntimeType) null)
    return (Type) this.m_bakedRuntimeType;
if (!this.IsEnum)
    return (Type) this;
if (this.m_enumUnderlyingType == (Type) null)
    throw new ...;
return this.m_enumUnderlyingType;
```

So it looks like in the cases like issue #2272, either `this.IsEnum` is false (which doesn't make a lot of sense) or `m_enumUnderlyingType` is actually set to the enum type instead of the underlying type. `m_enumUnderlyingType` looks like it only gets set during calls to `TypeBuilder.DefineField(...)`.

Is this making any sense? It's entirely possible I could just be rambling :). At this point I'll wait until at least tomorrow to do more digging in case someone has a suspicion what's going on.

**Comment by daveaglick** [comment]
That should do it. I ended up checking whether the enum was generated or defined out of scope and then getting the underlying type differently depending:
```
paramType = paramType is TypeBuilder ? paramType.UnderlyingSystemType : Enum.GetUnderlyingType(paramType);
```

This passes all existing tests as well as a new one I added for issue #2272 that failed with the original code.

**Comment by daveaglick** [comment]
Is there any feedback on this PR? Should I be looking somewhere else to fix the problem identified in issue #2272?

**Comment by KevinH-MS** [comment]
@tmat could you take a look?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs [hunk 1]
```diff
@@ -11436,6 +11436,30 @@ .maxstack  2
 }");
         }
 
+        [Fact]
+        public void Bug2272()
```
Comment (KevinH-MS) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTests.cs]: Consider:  use descriptive name for name of method (even if it's long) and use [WorkItem] attribute to denote the bug id.

**Comment by tmat** [comment]
:+1:

**Comment by KevinH-MS** [comment]
Thanks, Dave!

FYI, I'm going to merge this and rename the test/add WorkItem attribute.

**Comment by daveaglick** [comment]
Awesome, thanks!

**Pull request merged**